README
NAME

  DBD::ADO - A DBI driver for Microsoft ADO (Active Data Objects)

SYNOPSIS

  use DBI();

  my $dbh = DBI->connect("dbi:ADO:$dsn", $user, $passwd );

DESCRIPTION

  The DBD::ADO module supports ADO access on a Win32 machine. DBD::ADO is
  written to support the standard DBI interface to data sources.

PREREQUISITES

  DBI
  Win32::OLE

  ADO (http://www.microsoft.com/Data/download.htm)

INSTALLATION

  To install this module type the following:

     perl Makefile.PL
     make
     make test
     make install

  You need to use the correct make command. That may be nmake or dmake,
  depending on which development environment you are using.
  If using the Win32 ActiveState build then it is recommended that you
  use the ppm utility to fetch and install the module instead.

TESTING

  The supplied tests will connect to the database using the DBI
  environment variables, e.g.:

    set DBI_DSN=dbi:ADO:File Name=test.udl
    set DBI_DSN=dbi:ADO:Provider=Microsoft.Jet.OLEDB.4.0;Data Source=C:\tmp\test.mdb

  The tests will create tables (or other database objects) with the name
  (or prefix) 'PERL_DBD_TEST' in the current schema.
  Check for the unlikly case that this will cause unwanted side effects!

  To run the tests, type

    make test

  If all tests pass, you'll see something like:

    t\01base.......ok
    t\02ads........ok
    t\02cxn........ok
    t\04os.........ok
        25/45 skipped: various reasons
    ...
    t\51qi.........ok
    All tests successful, 53 subtests skipped.
    Files=29, Tests=737, 122 wallclock secs ...

  It is possible to run individual test scripts, e.g.:

    perl -w -Mblib t\02cxn.t

AUTHORS

  Tim Bunce and Phlip. With many thanks to Jan Dubois and Jochen Wiedmann
  for additions, debuggery and general help.
  Special thanks to Thomas Lowery, who maintained this module 2001-2003.
  Current maintainer is Steffen Goeldner.

SUPPORT

  This software is supported via the dbi-users mailing list.
  For more information and to keep informed about progress you can join the
  mailing list by sending a message to dbi-users-help@perl.org

  Please post details of any problems (or changes you needed to make) to
  dbi-users@perl.org and CC them to me (sgoeldner@cpan.org).

COPYRIGHT

  Copyright (c) 1998, Tim Bunce
  Copyright (c) 1999, Tim Bunce, Phlip, Thomas Lowery
  Copyright (c) 2000, Tim Bunce, Thomas Lowery
  Copyright (c) 2001, Tim Bunce, Thomas Lowery, Steffen Goeldner
  Copyright (c) 2002, Thomas Lowery, Steffen Goeldner
  Copyright (c) 2003, Thomas Lowery, Steffen Goeldner
  Copyright (c) 2004-2009 Steffen Goeldner

  All rights reserved.

  You may distribute under the terms of either the GNU General Public
  License or the Artistic License, as specified in the Perl README file.


Changes

2017-01-24 2.99.3

  Patch to allow more_results on a stored procedure.

2021-06-30  2.99.2
  Hack to allow execute() to accept special @bind_values that are array reference and be un-referenced to be passed to bind_param.

2015-08-25  2.99.1
  Hack to force bind_param to always use a specific value length.

2011-03-22  2.99

  Patch to remove annoying warnings during global destruction,
  with tests, from Rafael Kitover.

  Implemented more_results().
  Added tests (t/26more.t).

  Small change to table_info(): no need for client-side cursors.

2011-02-04  2.98

  Fixed a 'Missing argument in sprintf at DBD/ADO/GetInfo.pm line 13'.

  Patch for supporting cygwin in Makefile.PL from Rafael Kitover.

  Added a workaround for ADO Providers that don't accept a parameter size 0,
  spotted by Jeff Picklyk.
  Added test (12bind.t).

  Added the CLONE special subroutine, spotted by Jeff Picklyk.

2009-03-22  2.97

  Added tests for bind_param_inout() (t/28sp.t).

  Patch for supporting bind_param_inout() from Peter Mei.

2006-12-17  2.96

  Implemented statistics_info().
  Added tests (t/45ddst.t).

  Fixed bug: Application uses a value of the wrong type for the current
  operation, reported by amonotod.

  Fixed bug: Can't use string ("3") as a HASH ref while "strict refs"
  in use at ADO.pm line 1022, reported by John Derrick.

  Added support for the adExecuteStream option thanks to Luke Bakken.

2005-12-04  2.95

  Changed bind_param() to use an optional ado_size attribute.
  This is a workaround for the SQLOLEDB bug documented earlier.

  Replaced global error variables with handle attributes.

2005-01-30  2.94

  Reworked db/st handle attributes:
  - fetching all the data of a SELECT sets Active off
  - dropped the RowsInCache attribute (ADO doesn't provide this
    information and it was confused with RowCacheSize anyway)
  - added support for the RowCacheSize attribute
  - dropped the User attribute
  - deprecated CursorType, Filter, ADO_Columns, Trim_Catalog
    (use the corresponding ado_* attributes)
  - added ado_ConnectionTimeout and ado_Mode to avoid the abuse of the
    ConnectionString

  Renamed ADOTEST to DBD_TEST.

2004-10-17  2.93

  Implemented do().

  Changed $sth->DESTROY to call finish( $sth ) instead of $sth->finish.
  The old call caused clearing of errors e.g. in $dbh->do().

  Replaced duplicated code in disconnect() with a call to rollback().

  Minor changes to db/st handle attributes.

  Minor changes to table_info()'s parameter handling.

  Modified tests (t/19misc.t, t/25curs.t) to use bind_param().

  Added INSERT tests (t/12ins_b.t, t/12ins_q.t).

  Documented a SQLOLEDB bug, reported by Vahe Sarkissian.

2004-09-19  2.92

  Enabled and reimplemented LOB tests (t/27lob.t, t/19misc.t).

  Added support for LongReadLen/LongTruncOk.

  Minor changes to the STORE() methods.

  Moved more datatype related stuff into DBD::ADO::TypeInfo.

  Implemented some timeout tests (t/29timeout.t).

  Changed a test in t/14rows.t due to an API change in
  Test::More 0.34 (is()'s undef/'' handling).

  Removed the $sth->{BadAttributeHere} test from t/21sth.t,
  because getting/setting an invalid attribute no longer is
  a fatal error in DBI 1.43.

2004-05-02  2.91

  Adjusted the trace levels in trace_msg().

  Removed disconnect_all().

2004-03-21  2.90

  Changed default for $dbh->{ado_ti_ver} to 2.

2004-02-15  2.89

  Added an alternative implementation for type_info_all(), which is
  enabled if $dbh->{ado_ti_ver} == 2.

2004-02-08  2.88

  Removed $dbh->{ado_provider_auto_commit_comments} and renamed
  $dbh->{ado_provider_support_auto_commit} to $dbh->{ado_txn_capable}.

  Changed $dbh->FETCH to handle only 'ado_'-attributes.

  Replaced duplicated code in DESTROY() with a call to finish().

2004-02-01  2.87

  Fixed a bug when fetching data from a result set with non-unique
  column names, reported by Ed Belisle.

  Removed DBD::ADO::dr::DESTROY.

2004-01-25  2.86

  Added support for $sth->{ParamValues}.

  Added attribute $h->{ado_max_errors} to avoid time-consuming processing
  of huge error collections, suggested by Peter Hircock.

  Removed private attributes from database and statement handles.

2004-01-18  2.85

  Improved error handling to avoid problems with success/informational
  messages, reported by Ed Belisle and Peter Hircock.
  These 'success with information' conditions no longer trigger the normal
  DBI error handling mechanisms.

2004-01-11  2.84

  Reimplemented NULL tests (t/23null.t).

  Fixed a DBD::Proxy related bug, reported by Peter Hircock.
  The execute() method returned a Win32::OLE::Variant object which
  couldn't be restored at the client w/o Win32::OLE.

2003-12-14  2.83

  Added notes about Unicode support thanks to Stefan Ram and Jan Dubois.

  Rearranged some tests.

  Installed ADO-specific method ado_open_schema().

2003-12-07  2.82

  Replaced duplicated code in DESTROY() with a call to disconnect().

  Removed the deprecated Enums() method from DBD::ADO::TypeInfo.

  Added t/02ads.t to test data_sources() (and removed a similar test
  from t/02simple.t).

  Removed type_info() tests from t/17conn.t.

2003-11-30  2.81

  Fixed Makefile.PL to work with strict 1.03 (perl 5.8.1.),
  reported by Eric Thinnes.

  Added ping().
  Added tests for ping() to t/02cxn.t.
  Removed ping() tests from t/15func.t.

  Fixed connect() to set 'Active' on.
  Added a test for the 'Active' flag to t/02cxn.t

  Moved OpenSchema() tests from t/11curs.t into t/04os.t.

2003-11-23  2.80

  Fixed a bug affecting OpenSchema (first row was fetched twice).

  Improved the type handling of bind_param().

2003-11-16  2.79

  Added t/51qi.t to test quoted identifiers (and removed similar
  tests from t/15func.t).


Makefile.PL
#!/usr/local/bin/perl -sw
#
# You may distribute under the terms of either the GNU General Public
# License or the Artistic License, as specified in the Perl README file.
#
BEGIN { require 5.004 }  # 5.004 is required for Win32

use Config;
use ExtUtils::MakeMaker 5.16, qw(&WriteMakefile $Verbose);
use Getopt::Long;
use DBI 1.20;  # must be installed first
use DBI::DBD;
use strict;

my %opts =
(
  NAME => 'DBD::ADO',
  VERSION_FROM => 'lib/DBD/ADO.pm',
  clean => {},
  dist => {
    DIST_DEFAULT => 'clean distcheck tardist',
    PREOP    => '$(MAKE) -f Makefile.old distdir',
    COMPRESS => 'gzip -v9',
    SUFFIX => 'gz'
  },
);
if ( $ExtUtils::MakeMaker::VERSION >= 5.43 ) {
  $opts{AUTHOR} = 'Tim Bunce, Phlip, Thomas Lowery and Steffen Goeldner ( http://dbi.perl.org )';
  $opts{ABSTRACT} = 'ADO driver for the DBI module.';
  $opts{PREREQ_PM} = { DBI => 0, 'Win32::OLE' => 0.16 };
  $opts{CAPI} = 'TRUE' if $Config{archname} =~ /-object\b/i;
}

$::opt_g = 0;

GetOptions('g!') or die 'Invalid arguments';

$opts{OPTIMIZE} = '-g -O0' if $::opt_g;

die "Currently DBD::ADO only works on Win32 machines!\n"
  if $^O !~ /win32|cygwin/i && $ENV{USER} !~ /tlowery/i;

print <<'TEXT';

Configuring DBD::ADO ...

>>> Remember to actually *READ* the README file!
>>> And re-read it if you have any problems.

TEXT

my $dbi_dir      = dbd_dbi_dir();
my $dbi_arch_dir = dbd_dbi_arch_dir();

WriteMakefile( %opts );

local $^W = 0;

print <<"TEXT";

The DBD::ADO tests will use these values for the database connection:
    DBI_DSN=$ENV{DBI_DSN}		e.g. dbi:ADO:demo
    DBI_USER=$ENV{DBI_USER}
    DBI_PASS=$ENV{DBI_PASS}

TEXT

warn "*** Warning: Not all required environment variables are set.  Using default.\n"
  unless $ENV{DBI_DSN} && $ENV{DBI_USER} && $ENV{DBI_PASS};

warn "*** Warning: DBI_DSN ($ENV{DBI_DSN}) doesn't start with 'dbi:ADO:'\n"
  if $ENV{DBI_DSN} && $ENV{DBI_DSN} !~ m/^dbi:ADO:/;

print "\n";

sub MY::postamble {
  return dbd_postamble( @_ );
}


lib/DBD/ADO/GetInfo.pm
package DBD::ADO::GetInfo;

use strict;

my $fmt = '%02d.%02d.%1d%1d%1d%1d';       # ODBC version string: ##.##.#####

my $sql_driver_ver = sprintf $fmt, split(/[\._]/,"$DBD::ADO::VERSION.0.0.0.0.0");

our %info =
(
     41 => \&sql_catalog_name_separator   # SQL_CATALOG_NAME_SEPARATOR
,    22 => \&sql_concat_null_behavior     # SQL_CONCAT_NULL_BEHAVIOR
,     6 =>  $INC{'DBD/ADO.pm'}            # SQL_DRIVER_NAME               # XXX
,     7 =>  $sql_driver_ver               # SQL_DRIVER_VER                # XXX
,    28 => \&sql_identifier_case          # SQL_IDENTIFIER_CASE
,    29 => \&sql_identifier_quote_char    # SQL_IDENTIFIER_QUOTE_CHAR
,    89 => \&sql_keywords                 # SQL_KEYWORDS
);

our %odbc2ado =
(
    114 => 'Catalog Location'             # SQL_CATALOG_LOCATION
,    42 => 'Catalog Term'                 # SQL_CATALOG_TERM
,     2 => 'Data Source Name'             # SQL_DATA_SOURCE_NAME
,    17 => 'DBMS Name'                    # SQL_DBMS_NAME
,    18 => 'DBMS Version'                 # SQL_DBMS_VERSION
#     6 => 'Provider Name'                # SQL_DRIVER_NAME               # XXX
#     7 => 'Provider Version'             # SQL_DRIVER_VER                # XXX
,    40 => 'Procedure Term'               # SQL_PROCEDURE_TERM
,    39 => 'Schema Term'                  # SQL_SCHEMA_TERM
,    45 => 'Table Term'                   # SQL_TABLE_TERM
,    47 => 'User Name'                    # SQL_USER_NAME
);

sub sql_catalog_name_separator
{
  my $dbh = shift;
  DBD::ADO::db::ado_schema_dbinfo_literal( $dbh,'CATALOG_SEPARATOR') ||'.';
}
sub sql_concat_null_behavior
{
  { 1 => 0 # SQL_CB_NULL
  , 2 => 1 # SQL_CB_NON_NULL
  }->{$_[0]->{ado_conn}->Properties->{'NULL Concatenation Behavior'}{Value}};
}
sub sql_identifier_case
{
  { 1 => 1 # SQL_IC_UPPER
  , 2 => 2 # SQL_IC_LOWER
  , 4 => 3 # SQL_IC_SENSITIVE
  , 8 => 4 # SQL_IC_MIXED
  }->{$_[0]->{ado_conn}->Properties->{'Identifier Case Sensitivity'}{Value}};
}
sub sql_identifier_quote_char
{
  my $dbh = shift;
  DBD::ADO::db::ado_schema_dbinfo_literal( $dbh,'QUOTE') ||
  DBD::ADO::db::ado_schema_dbinfo_literal( $dbh,'QUOTE_PREFIX') ||'"';
}
sub sql_keywords
{
  my $dbh = shift;
  my $sth = $dbh->func('adSchemaDBInfoKeywords','OpenSchema');
  my @Keywords = ();
  while ( my $row = $sth->fetch ) {
    push @Keywords, $row->[0];
  }
  return join ',', @Keywords;
}

1;
